selfdrive/ui/qt/onroad/model.cc

```diff
@@ -0,0 +1,194 @@
+    bg.setColorAt(0.5, QColor::fromHslF(112 / 360.0, 1.0, 0.68, 0.35));
+    bg.setColorAt(1.0, QColor::fromHslF(112 / 360.0, 1.0, 0.68, 0.0));
+  }
+
+  painter.setBrush(bg);
+  painter.drawPolygon(path_polygon);
+}
+
+void ModelRenderer::drawLead(QPainter &painter, const cereal::RadarState::LeadData::Reader &lead_data,
+                             const cereal::XYZTData::Reader &model_position, const QRect &surface_rect) {
+  const float d_rel = lead_data.getDRel();
+  const int z_index = get_path_length_idx(model_position, d_rel);
+  const QPointF lead_pos = mapToScreen(d_rel, -lead_data.getYRel(), model_position.getZ()[z_index] + 1.22f);
+
+  const float sz = std::clamp((25 * 30) / (d_rel / 3 + 30), 15.0f, 30.0f) * 2.35f;
+  const float x = std::clamp<float>(lead_pos.x(), 0.f, surface_rect.width() - sz / 2.0f);
+  const float y = std::min<float>(lead_pos.y(), surface_rect.height() - sz * 0.6f);
+
+  const float g_xo = sz / 5.0f;
+  const float g_yo = sz / 10.0f;
+
+  QPointF glow[] = {{x + (sz * 1.35) + g_xo, y + sz + g_yo}, {x, y - g_yo}, {x - (sz * 1.35) - g_xo, y + sz + g_yo}};
+  painter.setBrush(QColor(218, 202, 37, 255));
+  painter.drawPolygon(glow, std::size(glow));
+
+  QPointF chevron[] = {{x + (sz * 1.25), y + sz}, {x, y}, {x - (sz * 1.25), y + sz}};
+  const float alpha = calculateLeadAlpha(d_rel, lead_data.getVRel());
+  painter.setBrush(QColor(201, 34, 49, alpha));
+  painter.drawPolygon(chevron, std::size(chevron));
+}
+
+uint8_t ModelRenderer::calculateLeadAlpha(float d_rel, float v_rel) {
+  const float speed_threshold = 10.0f;
+  const float lead_threshold = 40.0f;
+  float alpha = 0.0f;
+  if (d_rel < lead_threshold) {
+    alpha = 255 * (1.0f - (d_rel / lead_threshold));
+    if (v_rel < 0) {
+      alpha += 255 * (-v_rel / speed_threshold);
+    }
+  }
+  return std::clamp<uint8_t>(alpha, 0, 255);
+}
+
+QPointF ModelRenderer::mapToScreen(float in_x, float in_y, float in_z) {
+  Eigen::Vector3f input(in_x, in_y, in_z);
+  auto pt = car_space_transform * input;
+  return {pt.x() / pt.z(), pt.y() / pt.z()};
+}
+
+void ModelRenderer::mapLineToPolygon(const cereal::XYZTData::Reader &line, float y_off, float z_off,
+                                     QPolygonF *pvd, int max_idx, bool allow_invert) {
+  const auto &x_coords = line.getX(), &y_coords = line.getY(), &z_coords = line.getZ();
+  pvd->clear();
+
+  for (int i = 0; i <= max_idx; ++i) {
+    float x = x_coords[i], y = y_coords[i], z = z_coords[i];
+    // highly negative x positions  are drawn above the frame and cause flickering, clip to zy plane of camera
+    if (x < 0) continue;
+
+    const QPointF left = mapToScreen(x, y - y_off, z + z_off);
+    const QPointF right = mapToScreen(x, y + y_off, z + z_off);
+    if (clip_region.contains(left) && clip_region.contains(right)) {
+      // For wider lines the drawn polygon will "invert" when going over a hill and cause artifacts
+      if (!allow_invert && !pvd->empty() && left.y() > pvd->back().y()) {
```

Comment:
Changes like this  (`pvd->size()` to `!pvd->empty()`) should go in a different PR since they are hard to spot and review when so many things are moved around.